core/server/master/src/main/java/alluxio/master/file/FileSystemJournalEntryMerger.java

```diff
@@ -70,6 +70,14 @@ else if (
             MutableInodeDirectory.fromJournalEntry(existingEntry.getInodeDirectory());
         if (entry.hasUpdateInode()) {
           inodeDirectory.updateFromEntry(entry.getUpdateInode());
+          // Update Inode directory does not contain directory fingerprint,
+          // so we still need to add the new inode journal entry to the list to keep the
+          // fingerprint update,
+          // while we still merge it with the existing inode directory on as best efforts.
+          if (entry.getUpdateInode().hasUfsFingerprint()
+              && !entry.getUpdateInode().getUfsFingerprint().equals("")) {
+            mJournalEntries.add(entry);
+          }
```

Comment:
i wonder if it is possible to achieve the same with proto merging logic, like we have this kind of code
```
CreateDirectoryPOptions mergedOptions = FileSystemOptionsUtils.createDirectoryDefaults(
    mFsContext.getPathConf(path)).toBuilder().mergeFrom(options).build();
```
Is it possible that when Inode entries are merged, they keep the ufs fingerprint?

If the above is not easily achievable, i'm fine with the current approach. 

But note that the correctness now almost fully rely on the UTs. With so many if-else branches i don't think anyone can quickly understand each branch and what are/are not covered. We need good UTs to guarantee the behavior and can almost be certain that if a behavior is not found in UT, it is undefined.

Reply:
the issue is that InodeDirectory journal entry protobuf does not contain a "fingerprint" field. So no matter how you merges it with others, we are not able to let it carry the fingerprint information, unless we update the protobuf, which is a too breaking change that we want to avoid.